lib/ansible/modules/cloud/amazon/aws_acm.py

```diff
@@ -0,0 +1,396 @@
+
+'''
+
+RETURN = '''
+certificate:
+  description: Information about the certificate which was uploaded
+  type: complex
+  returned: when C(state) is C(present)
+  contains:
+    arn:
+      description: The ARN of the certificate in ACM
+      type: str
+      returned: when C(state) is C(present)
+      sample: "arn:aws:acm:ap-southeast-2:123456789012:certificate/01234567-abcd-abcd-abcd-012345678901"
+    domain_name:
+      description: The domain name encoded within the public certificate
+      type: str
+      returned: when C(state) is C(present)
+      sample: acm.ansible.com
+arns:
+  description: A list of the ARNs of the certificates in ACM which were deleted
+  type: list
+  returned: when C(state) is C(absent)
+  sample:
+   - "arn:aws:acm:ap-southeast-2:123456789012:certificate/01234567-abcd-abcd-abcd-012345678901"
+'''
+
+
+from ansible.module_utils.aws.core import AnsibleAWSModule
+from ansible.module_utils.ec2 import boto3_conn, ec2_argument_spec, get_aws_connection_info
+from ansible.module_utils.aws.acm import ACMServiceManager
+from ansible.module_utils.crypto import get_fingerprint_from_pem_cert as get_fingerprint
+from ansible.module_utils._text import to_bytes, to_text
+from ssl import PEM_cert_to_DER_cert
+
+import re  # regex library
+
+
+# Takes in two text arguments
+# Each a PEM encoded certificate
+# Or a chain of PEM encoded certificates
+# May include some lines between each chain in the cert, e.g. "Subject: ..."
+# Returns True iff the chains/certs are functionally identical (including chain order)
+def chain_compare(a, b):
+
+    # Use regex to split up a chain or single cert into an array of base64 encoded data
+    # Using "-----BEGIN CERTIFICATE-----" and "----END CERTIFICATE----"
+    # Noting that some chains have non-pem data in between each cert
+    expr = re.compile(r"-+BEGIN\s+CERTIFICATE-+([a-zA-Z0-9\+\/=\s]+)-+END\s+CERTIFICATE-+")
+    chain_a_pem = pem_chain_split(a)
+    chain_b_pem = pem_chain_split(b)
+
+    if len(chain_a_pem) != len(chain_b_pem):
+        return(False)
+
+    # Chain length is the same
+    for (ca, cb) in zip(chain_a_pem, chain_b_pem):
+        der_a = PEM_body_to_DER(ca)
+        der_b = PEM_body_to_DER(cb)
+        if der_a != der_b:
+            return(False)
+
+    return(True)
+
+
+# Takes in PEM encoded data with no headers
+# returns equivilent DER
+def PEM_body_to_DER(pem):
+    prefix = "-----BEGIN CERTIFICATE-----"
+    suffix = "-----END CERTIFICATE-----"
+    pem_full = '\n'.join([prefix, pem, suffix])
+    return(PEM_cert_to_DER_cert(pem_full))
+
+
+# Use regex to split up a chain or single cert into an array of base64 encoded data
+# Using "-----BEGIN CERTIFICATE-----" and "----END CERTIFICATE----"
+# Noting that some chains have non-pem data in between each cert
+# This function returns only what's between the headers, excluding the headers
+def pem_chain_split(pem):
+    expr = re.compile(r"-+BEGIN\s+CERTIFICATE-+([a-zA-Z0-9\+\/=\s]+?)-+END\s+CERTIFICATE-+")
```

Comment:
Better store that in a global variable, otherwise there's no need to compile it before using it.

Reply:
About the expression: I would change it to
```suggestion
    expr = re.compile(r"------?BEGIN [A-Z0-9. ]*CERTIFICATE------?([a-zA-Z0-9\+\/=\s]+)------?END [A-Z0-9. ]*CERTIFICATE------?")
```
That covers some more corner cases (https://tools.ietf.org/html/rfc7468#appendix-A).

(Also see https://stackoverflow.com/a/5356351)

Reply:
done